pkg/apis/pipeline/v1/pipeline_types.go

```diff
@@ -0,0 +1,561 @@
+				// report an error if the task name does not exist in the list of dag tasks
+				if !ptNames.Has(pt) {
+					errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("pipeline task %s is not defined in the pipeline", pt), fieldPath))
+				}
+			}
+		}
+	}
+	return errs
+}
+
+func (pt *PipelineTask) validateWorkspaces(workspaceNames sets.String) (errs *apis.FieldError) {
+	for i, ws := range pt.Workspaces {
+		if ws.Workspace == "" {
+			if !workspaceNames.Has(ws.Name) {
+				errs = errs.Also(apis.ErrInvalidValue(
+					fmt.Sprintf("pipeline task %q expects workspace with name %q but none exists in pipeline spec", pt.Name, ws.Name),
+					"",
+				).ViaFieldIndex("workspaces", i))
+			}
+		} else if !workspaceNames.Has(ws.Workspace) {
+			errs = errs.Also(apis.ErrInvalidValue(
+				fmt.Sprintf("pipeline task %q expects workspace with name %q but none exists in pipeline spec", pt.Name, ws.Workspace),
+				"",
+			).ViaFieldIndex("workspaces", i))
+		}
+	}
+	return errs
+}
+
+// TaskSpecMetadata returns the metadata of the PipelineTask's EmbeddedTask spec.
+func (pt *PipelineTask) TaskSpecMetadata() PipelineTaskMetadata {
+	return pt.TaskSpec.Metadata
+}
+
+// HashKey is the name of the PipelineTask, and is used as the key for this PipelineTask in the DAG
+func (pt PipelineTask) HashKey() string {
+	return pt.Name
+}
+
+// ValidateName checks whether the PipelineTask's name is a valid DNS label
+func (pt PipelineTask) ValidateName() *apis.FieldError {
+	if err := validation.IsDNS1123Label(pt.Name); len(err) > 0 {
+		return &apis.FieldError{
+			Message: fmt.Sprintf("invalid value %q", pt.Name),
+			Paths:   []string{"name"},
+			Details: "Pipeline Task name must be a valid DNS Label." +
+				"For more info refer to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names",
+		}
+	}
+	return nil
+}
+
+// Validate classifies whether a task is a custom task or a regular task(dag/final)
+// calls the validation routine based on the type of the task
+func (pt PipelineTask) Validate(ctx context.Context) (errs *apis.FieldError) {
+	errs = errs.Also(pt.validateRefOrSpec())
+
+	errs = errs.Also(pt.validateEmbeddedOrType())
+
+	cfg := config.FromContextOrDefaults(ctx)
+	// If EnableCustomTasks feature flag is on, validate custom task specifications
+	// pipeline task having taskRef with APIVersion is classified as custom task
+	switch {
+	case cfg.FeatureFlags.EnableCustomTasks && pt.TaskRef != nil && pt.TaskRef.APIVersion != "":
+		errs = errs.Also(pt.validateCustomTask())
+	case cfg.FeatureFlags.EnableCustomTasks && pt.TaskSpec != nil && pt.TaskSpec.APIVersion != "":
+		errs = errs.Also(pt.validateCustomTask())
+	default:
+		errs = errs.Also(pt.validateTask(ctx))
+	}
+	return
+}
+
+// Deps returns all other PipelineTask dependencies of this PipelineTask, based on ordering
+func (pt PipelineTask) Deps() []string {
+	deps := []string{}
+
+	deps = append(deps, pt.orderingDeps()...)
```

Comment:
It seems to me that the removal of resource dependencies was a bit too eager here.
We do not need in `v1` dependencies related to `PipelineResources` but we still do need dependencies associated with `results` references.